Hibernate-Validator: Custom `ValueExtractor` is not automatically registered

### Describe the bug

A custom `ValueExtractor` is not automatically registered by quarkus (https://quarkus.io/guides/validation#hibernate-validator-extension-and-cdi).

```bash
2021-09-23 08:13:01,190 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
2021-09-23 08:13:01,193 ERROR [io.qua.run.boo.StartupActionImpl] (Quarkus Main Thread) Error running Quarkus: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:98)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:65)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:119)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
	... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:428)
	... 15 more
Caused by: javax.validation.ConstraintDeclarationException: HV000197: No value extractor found for type parameter 'T' of type org.acme.Container.
	at org.hibernate.validator.internal.metadata.core.MetaConstraints.addValueExtractorDescriptorForTypeArgumentLocation(MetaConstraints.java:147)
	at org.hibernate.validator.internal.metadata.core.MetaConstraints.create(MetaConstraints.java:63)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.findTypeUseConstraints(AnnotationMetaDataProvider.java:805)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.findTypeArgumentsConstraints(AnnotationMetaDataProvider.java:775)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.findTypeArgumentsConstraints(AnnotationMetaDataProvider.java:778)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.findTypeAnnotationConstraintsForExecutableParameter(AnnotationMetaDataProvider.java:729)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.getParameterMetaData(AnnotationMetaDataProvider.java:431)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.findExecutableMetaData(AnnotationMetaDataProvider.java:308)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.getMetaData(AnnotationMetaDataProvider.java:292)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.getMethodMetaData(AnnotationMetaDataProvider.java:279)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.retrieveBeanConfiguration(AnnotationMetaDataProvider.java:130)
	at org.hibernate.validator.internal.metadata.provider.AnnotationMetaDataProvider.getBeanConfiguration(AnnotationMetaDataProvider.java:120)
	at org.hibernate.validator.internal.metadata.PredefinedScopeBeanMetaDataManager.getBeanConfigurationForHierarchy(PredefinedScopeBeanMetaDataManager.java:183)
	at org.hibernate.validator.internal.metadata.PredefinedScopeBeanMetaDataManager.createBeanMetaData(PredefinedScopeBeanMetaDataManager.java:150)
	at org.hibernate.validator.internal.metadata.PredefinedScopeBeanMetaDataManager.<init>(PredefinedScopeBeanMetaDataManager.java:100)
	at org.hibernate.validator.internal.engine.PredefinedScopeValidatorFactoryImpl.<init>(PredefinedScopeValidatorFactoryImpl.java:180)
	at org.hibernate.validator.PredefinedScopeHibernateValidator.buildValidatorFactory(PredefinedScopeHibernateValidator.java:42)
	at org.hibernate.validator.internal.engine.AbstractConfigurationImpl.buildValidatorFactory(AbstractConfigurationImpl.java:433)
	at io.quarkus.hibernate.validator.runtime.HibernateValidatorRecorder$1.created(HibernateValidatorRecorder.java:139)
	at io.quarkus.arc.runtime.ArcRecorder.initBeanContainer(ArcRecorder.java:70)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy_0(ArcProcessor$generateResources-1025303321.zig:139)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy(ArcProcessor$generateResources-1025303321.zig:40)
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:386)
	... 15 more

2021-09-23 08:13:01,264 INFO  [org.jbo.threads] (main) JBoss Threads version 3.4.2.Final
```

Note: Registration via Service-Loader-API works as expected (https://docs.jboss.org/hibernate/validator/6.2/reference/en-US/html_single/#section-valueextraction-registeringvalueextractor).

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

1. https://code.quarkus.io/?e=resteasy-reactive-jackson&e=hibernate-validator
2. Add the following classes:

```java
package org.acme;

public class Container<T> {
}
```

```java
package org.acme;

import javax.enterprise.context.ApplicationScoped;
import javax.validation.valueextraction.ExtractedValue;
import javax.validation.valueextraction.ValueExtractor;

@ApplicationScoped
public class ContainerValueExtractor
        implements ValueExtractor<Container<@ExtractedValue ?>> {

    @Override
    public void extractValues(Container<?> originalValue, ValueReceiver receiver) {
        // whatever
    }
}
```

```java
package org.acme;

import javax.enterprise.context.ApplicationScoped;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import java.util.List;

@ApplicationScoped
@Path("/container")
public class ContainerResource {

    @GET
    @Path("/1")
    @Produces(MediaType.TEXT_PLAIN)
    public String containerMethod(
            @Valid List<@NotNull @NotNull Container<@NotNull String>> list
    ) {
        return "hello";
    }
}
```

3. `./mvnw quarkus:dev`

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

Java 11

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

Quarkus 2.2.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.1

### Additional information

_No response_

> 
> 
> > In the reproducer above there isn't an external module. So it should work out of the box.
> 
> Not really a reproducer if it doesn't reproduce the problem ;)

Ok, i checked it. I followed exactly my description to reproduce the problem and i can reproduce the issue without problems. Even a simple `./mvnw clean install` leads to the following error:

```bash
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 7.075 s <<< FAILURE! - in org.acme.ReactiveGreetingResourceTest
[ERROR] org.acme.ReactiveGreetingResourceTest.testHelloEndpoint  Time elapsed: 0.008 s  <<< ERROR!
java.lang.RuntimeException: java.lang.ExceptionInInitializerError
Caused by: java.lang.ExceptionInInitializerError
Caused by: java.lang.RuntimeException: Failed to start quarkus
Caused by: javax.validation.ConstraintDeclarationException: HV000197: No value extractor found for type parameter 'T' of type org.acme.Container.

[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors: 
[ERROR]   ReactiveGreetingResourceTest.testHelloEndpoint » Runtime java.lang.ExceptionIn...
[INFO]
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0
```

> 
> > Do we have to create an empty file META-INF/beans.xml?
> 
> According to [the Quarkus CDI documentation](https://quarkus.io/guides/cdi-reference#bean_discovery), having a Jandex index in your dependency at `META-INF/jandex.idx` should be enough.
> 
> But adding `META-INF/beans.xml` is worth a try; at least it will give us more information about what's wrong exactly. Please try?
> 

I tried out your suggestion. Adding a `META-INF/beans.xml` doesn't help.

> Or provide a reproducer, then we can try things ourselves :)

See above.

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/ApplicationScopedCustomValueExtractorTest.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/ApplicationScopedCustomValueExtractorTest.java
@@ -1,0 +1,57 @@
+package io.quarkus.hibernate.validator.test.valueextractor;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.validation.ValidatorFactory;
+import javax.validation.constraints.NotBlank;
+import javax.validation.valueextraction.ExtractedValue;
+import javax.validation.valueextraction.ValueExtractor;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+@Disabled("Reproduces https://github.com/quarkusio/quarkus/issues/20375, not yet fixed")
+public class ApplicationScopedCustomValueExtractorTest {
+
+    @Inject
+    ValidatorFactory validatorFactory;
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(TestBean.class, Container.class, ApplicationScopedContainerValueExtractor.class));
+
+    @Test
+    public void testApplicationScopedCustomValueExtractor() {
+        assertThat(validatorFactory.getValidator().validate(new TestBean())).hasSize(1);
+    }
+
+    public static class TestBean {
+        public Container<@NotBlank String> constrainedContainer;
+
+        public TestBean() {
+            Container<String> invalidContainer = new Container<>();
+            invalidContainer.value = "   ";
+
+            this.constrainedContainer = invalidContainer;
+        }
+    }
+
+    @ApplicationScoped
+    public static class ApplicationScopedContainerValueExtractor
+            implements ValueExtractor<Container<@ExtractedValue ?>> {
+
+        @Override
+        public void extractValues(Container<?> originalValue, ValueReceiver receiver) {
+            receiver.value("someName", originalValue.value);
+        }
+    }
+
+}

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/Container.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/Container.java
@@ -2,6 +2,7 @@ package io.quarkus.hibernate.validator.test.valueextractor;
 
 // TODO for some reason, this must be a top-level type, not a nested/inner type,
 //  in order for annotations on type parameters to be detected.
+//  See NestedContainerTypeCustomValueExtractorTest
 public class Container<T> {
 
     public T value;

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/Container.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/Container.java
@@ -1,0 +1,9 @@
+package io.quarkus.hibernate.validator.test.valueextractor;
+
+// TODO for some reason, this must be a top-level type, not a nested/inner type,
+//  in order for annotations on type parameters to be detected.
+public class Container<T> {
+
+    public T value;
+
+}

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/NestedContainerTypeCustomValueExtractorTest.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/NestedContainerTypeCustomValueExtractorTest.java
@@ -1,0 +1,62 @@
+package io.quarkus.hibernate.validator.test.valueextractor;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.validation.ValidatorFactory;
+import javax.validation.constraints.NotBlank;
+import javax.validation.valueextraction.ExtractedValue;
+import javax.validation.valueextraction.ValueExtractor;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+@Disabled("Reproduces https://github.com/quarkusio/quarkus/issues/20377, not yet fixed")
+public class NestedContainerTypeCustomValueExtractorTest {
+
+    @Inject
+    ValidatorFactory validatorFactory;
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(TestBean.class, NestedContainerType.class, NestedContainerClassValueExtractor.class));
+
+    @Test
+    public void testNestedContainerTypeValueExtractor() {
+        assertThat(validatorFactory.getValidator().validate(new TestBean())).hasSize(1);
+    }
+
+    public static class TestBean {
+        public NestedContainerType<@NotBlank String> constrainedContainer;
+
+        public TestBean() {
+            NestedContainerType<String> invalidContainer = new NestedContainerType<>();
+            invalidContainer.value = "   ";
+
+            this.constrainedContainer = invalidContainer;
+        }
+    }
+
+    public static class NestedContainerType<T> {
+
+        public T value;
+
+    }
+
+    @Singleton
+    public static class NestedContainerClassValueExtractor
+            implements ValueExtractor<NestedContainerType<@ExtractedValue ?>> {
+
+        @Override
+        public void extractValues(NestedContainerType<?> originalValue, ValueReceiver receiver) {
+            receiver.value("someName", originalValue.value);
+        }
+    }
+}

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/SingletonCustomValueExtractorTest.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/valueextractor/SingletonCustomValueExtractorTest.java
@@ -1,0 +1,55 @@
+package io.quarkus.hibernate.validator.test.valueextractor;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.validation.ValidatorFactory;
+import javax.validation.constraints.NotBlank;
+import javax.validation.valueextraction.ExtractedValue;
+import javax.validation.valueextraction.ValueExtractor;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class SingletonCustomValueExtractorTest {
+
+    @Inject
+    ValidatorFactory validatorFactory;
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(TestBean.class, Container.class, SingletonContainerValueExtractor.class));
+
+    @Test
+    public void testSingletonCustomValueExtractor() {
+        assertThat(validatorFactory.getValidator().validate(new TestBean())).hasSize(1);
+    }
+
+    public static class TestBean {
+        public Container<@NotBlank String> constrainedContainer;
+
+        public TestBean() {
+            Container<String> invalidContainer = new Container<>();
+            invalidContainer.value = "   ";
+
+            this.constrainedContainer = invalidContainer;
+        }
+    }
+
+    @Singleton
+    public static class SingletonContainerValueExtractor
+            implements ValueExtractor<Container<@ExtractedValue ?>> {
+
+        @Override
+        public void extractValues(Container<?> originalValue, ValueReceiver receiver) {
+            receiver.value("someName", originalValue.value);
+        }
+    }
+
+}

--- a/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorRecorder.java
+++ b/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorRecorder.java
@@ -3,6 +3,7 @@ package io.quarkus.hibernate.validator.runtime;
 import java.util.Set;
 import java.util.function.Supplier;
 
+import javax.enterprise.util.TypeLiteral;
 import javax.validation.ClockProvider;
 import javax.validation.ConstraintValidatorFactory;
 import javax.validation.MessageInterpolator;
@@ -31,6 +32,9 @@ import io.quarkus.runtime.annotations.Recorder;
 
 @Recorder
 public class HibernateValidatorRecorder {
+
+    private static final TypeLiteral<ValueExtractor<?>> TYPE_LITERAL_VALUE_EXTRACTOR_WITH_WILDCARD = new TypeLiteral<ValueExtractor<?>>() {
+    };
 
     public BeanContainerListener initializeValidatorFactory(Set<Class<?>> classesToBeValidated,
             Set<String> detectedBuiltinConstraints,
@@ -132,7 +136,10 @@ public class HibernateValidatorRecorder {
 
                 // Automatically add all the values extractors declared as beans
                 for (ValueExtractor<?> valueExtractor : Arc.container().beanManager().createInstance()
-                        .select(ValueExtractor.class)) {
+                        // Apparently passing ValueExtractor.class
+                        // won't match beans implementing ValueExtractor<NotAWildcard>,
+                        // so we need a type literal here.
+                        .select(TYPE_LITERAL_VALUE_EXTRACTOR_WITH_WILDCARD)) {
                     configuration.addValueExtractor(valueExtractor);
                 }
 

